File: diagonal.java
// linear time o(m*n) and linear space
class Solution{
    public int[] diag(int[][] matrix){
        int m = matrix.length;
        int n = matrix[0].length;
        boolean dir = true;
        int []result = new int[m*n];
        int idx = 0;
        int i = 0; int j = 0;
        while(idx < m*n){
            result[idx] = matrix[m][n];
            idx++;
            if(dir){
                if(i==0 && j != n-1){
                    j++;
                    dir = false;

                else if(j == n){
                    i++;
                    dir = false;

                else{
                    i--;j++;
                }

                
                }
            else{
                if(j==0 && i != n-1){
                    i++;
                    dir = true;

                else if(i == m-1){
                    j++;
                    dir = true;

                else{
                    j--;i++;
                }

            }
            }

        }
        return result;

    }
}

File: Diagonal.py
# Time Complexity: O(m*n)
# Space Complexity: O(1)
  
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        
        m = len(mat)
        n = len(mat[0])
        
        result = [0 for i in range(m*n)]
        
        i = 0
        j = 0
        dir = 1
        idx = 0
        
        while idx < m*n:
            result[idx] = mat[i][j]
            idx += 1
            
            if dir == 1:
                if (i == 0 and j != n-1):
                    j += 1
                    dir = -1
                elif j == n-1:
                    i += 1
                    dir = -1
                else:
                    i -= 1
                    j += 1
            else:
                if (j == 0 and i != m-1):
                    i += 1
                    dir = 1
                elif i == m-1:
                    j += 1
                    dir = 1
                else:
                    i += 1
                    j -= 1
                    
         
        return result        
            
        
